lychee-lib/src/types/resolver.rs

```diff
@@ -0,0 +1,77 @@
+use super::{FileType, InputContent, InputSource};
+use crate::chain::Chain;
+use crate::utils::request;
+use crate::{BasicAuthExtractor, ChainResult, ErrorKind, Handler, Result, Uri};
+use async_trait::async_trait;
+use http::HeaderMap;
+use reqwest::{Client, Request, Url};
+
+/// Structure to fetch remote content.
+#[derive(Debug, Default, Clone)]
+pub struct UrlContentResolver {
+    pub basic_auth_extractor: Option<BasicAuthExtractor>,
+    pub headers: HeaderMap,
+    pub client: reqwest::Client,
+}
+
+type RequestChain = Chain<Request, Result<String>>;
+
+impl UrlContentResolver {
+    /// Fetch remote content by URL.
+    /// This method is not intended to check if a URL is functional but
+    /// to get a URL's content and process the content.
+    pub async fn url_contents(&self, url: Url) -> Result<InputContent> {
+        // Assume HTML for default paths
+        let file_type = if url.path().is_empty() || url.path() == "/" {
+            FileType::Html
+        } else {
+            FileType::from(url.as_str())
+        };
+
+        let credentials = request::extract_credentials(
+            self.basic_auth_extractor.as_ref(),
+            &Uri { url: url.clone() },
+        );
+
+        let chain: RequestChain = Chain::new(vec![Box::new(credentials), Box::new(self.clone())]);
+
+        let request = self
+            .client
+            .request(reqwest::Method::GET, url.clone())
+            .build()
+            .map_err(ErrorKind::BuildRequestClient)?;
+
+        let content = match chain.traverse(request).await {
+            ChainResult::Next(_) => unreachable!(
+                "ChainResult::Done is unconditionally returned from the last chain element"
+            ),
```

Comment:
That feels like a bit of a wart in our chain implementation.
I wonder how we could avoid that. 🤔 
Should `traverse` return the final `Result` instead? It's probably not possible, but I haven't thought much about it.

Reply:
Makes me wonder if we should even use a request chain here. After all, we just call `extract_credentials` and then `client.request`, right? The chain would be helpful for inversion of control like in situations where we want to return the chain and someone else "merges" it with other handlers and then calls the entire thing.

However, in this function here, the execution is very imperative as in "do this, then do that." 🤔 So maybe we just get rid of the request chain here? Or was there a specific reason for using it? The alternative would be to make the entire resolver a handler in the request chain. Then it would make sense.

Feedback welcome!